gnocchi/amqp1d.py

```diff
@@ -0,0 +1,251 @@
+
+from gnocchi import incoming
+from gnocchi import indexer
+from gnocchi import service
+
+from gnocchi import utils
+
+from proton.handlers import MessagingHandler
+from proton.reactor import Container
+
+LOG = daiquiri.getLogger(__name__)
+
+
+class CollectdStats(object):
+    def __init__(self, conf):
+        self.conf = conf
+        self.incoming = incoming.get_driver(self.conf)
+        self.indexer = indexer.get_driver(self.conf)
+        self.resource_type = None
+
+        self._ensure_resource_type_exists()
+        self.gauges = {}
+        self.counters = {}
+        self.absolute = {}
+
+    def reset(self):
+        self.gauges.clear()
+        self.counters.clear()
+        self.absolute.clear()
+
+    def on_timer_task(self, event):
+        event.container.schedule(self.conf.amqp1d.flush_delay, self)
+        self.flush()
+
+    @staticmethod
+    def serialize_identifier(index, value):
+        """Based of FORMAT_VL from collectd/src/daemon/common.h.
+
+        The biggest difference is that we don't prepend the host and append the
+        index of the value, and don't use slash.
+
+        """
+        return (value["plugin"] + ("-" + value["plugin_instance"]
+                                   if value["plugin_instance"] else "")
+                + "@"
+                + value["type"] + ("-" + value["type_instance"]
+                                   if value["type_instance"] else "")
+                + "-" + str(index))
+
+    def treat_metric(self, host, metric_name, metric_type,
+                     value):
+        """Collectd.
+
+        Statistics in collectd consist of a value list. A value list includes:
+        Values, can be one of:
+        Derive: used for values where a change in the value since it's last
+        been read is of interest. Can be used to calculate and store a rate.
+        Counter: similar to derive values, but take the possibility of a
+        counter wrap around into consideration.
+        Gauge: used for values that are stored as is.
+        Absolute: used for counters that are reset after reading.
+
+        """
+
+        if metric_type == "absolute":
+            if host not in self.absolute:
+                self.absolute[host] = {}
+            self.absolute[host][metric_name] = incoming.Measure(
+                utils.dt_in_unix_ns(utils.utcnow()), value)
+        elif metric_type == "gauge":
+            if host not in self.gauges:
+                self.gauges[host] = {}
+            self.gauges[host][metric_name] = incoming.Measure(
+                utils.dt_in_unix_ns(utils.utcnow()), value)
+        elif metric_type == "counter" or metric_type == "derive":
+            if host not in self.counters:
+                self.counters[host] = {}
+            self.counters[host][metric_name] = incoming.Measure(
+                utils.dt_in_unix_ns(utils.utcnow()), value)
+        else:
+            raise ValueError("Unknown metric type '%s'" % metric_type)
+
+    def flush(self):
```

Comment:
Is there any reason you use a `flush()` based model here?

It seems you treat all counters, gauge and absolute in the same way in the end, so I don't understand why you just don't write them as soon as you receive them?

Reply:
Sure, I will make changes to write to as I receive them. When I started I was not sure how to proceed with data types. Will have that updated.